cpp/src/arrow/testing/random.cc

```diff
@@ -558,5 +584,248 @@ std::shared_ptr<Array> RandomArrayGenerator::ArrayOf(std::shared_ptr<DataType> t
   return RandomArrayGeneratorOfImpl{this, type, size, null_probability, nullptr}.Finish();
 }
 
+namespace {
+template <typename T>
+typename T::c_type GetMetadata(const KeyValueMetadata* metadata, const std::string& key,
+                               typename T::c_type default_value) {
```

Comment:
This doesn't need an explicit template argument; it can infer from `default_value`
```suggestion
template <typename T, typename ArrowType = typename CTypeTraits<T>::ArrowType>
enable_if_parameter_free<ArrowType, T> GetMetadata(const KeyValueMetadata* metadata, const std::string& key, T default_value) {
```
(not necessary if you prefer to be explicit)

Reply:
Nice, I admit I'm not as up to speed on my template metaprogramming as I'd like to be.